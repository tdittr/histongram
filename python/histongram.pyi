from typing import Iterator


class Histogram:
    def __init__(self): ...

    def add(self, key: str): ...

    def add_many(self, iter: Iterator[str]): ...

    def __getitem__(self, key: str) -> int: ...
