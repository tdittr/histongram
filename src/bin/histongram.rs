use std::fs::read_to_string;
use std::path::PathBuf;

use clap::Parser;
use color_eyre::Result;
use compact_str::CompactString;

use histongram::{Histogram, Ngrams};

#[derive(Parser, Debug, Clone)]
struct Args {
    file: PathBuf,

    #[clap(long)]
    print: bool,
}

fn main() -> Result<()> {
    color_eyre::install()?;
    let args = Args::parse();

    let data = read_to_string(args.file)?;

    let ngrams = Ngrams::new(1..=5).count(data.split_whitespace());

    if args.print {
        println!("{ngrams:?}");
    }

    Ok(())
}
