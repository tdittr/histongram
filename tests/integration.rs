use std::cmp::{min, Reverse};

use insta::assert_ron_snapshot;

use histongram::{DefaultHashBuilder, Histogram};

const APACHE: &str = include_str!("../LICENSE-APACHE");
const MIT: &str = include_str!("../LICENSE-MIT");

#[test]
fn license_words() {
    let mut a: Histogram<&'static str> = APACHE.split_whitespace().collect();
    let mut a_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut a_counts);
    assert_ron_snapshot!(a_counts);

    let m: Histogram<_> = MIT.split_whitespace().collect();
    let mut m_counts = m.clone().sorted_occurrences();
    sort_also_by_key(&mut m_counts);
    assert_ron_snapshot!(m_counts);

    a.append(m);
    let mut combined_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut combined_counts);
    assert_ron_snapshot!(combined_counts);
}

#[test]
fn license_chars() {
    let mut a = Histogram::<_, DefaultHashBuilder>::from_owned_iter(APACHE.chars());
    let mut a_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut a_counts);
    assert_ron_snapshot!(a_counts);

    let m = Histogram::from_owned_iter(MIT.chars());
    let mut m_counts = m.clone().sorted_occurrences();
    sort_also_by_key(&mut m_counts);
    assert_ron_snapshot!(m_counts);

    a.append(m);
    let mut combined_counts = a.clone().sorted_occurrences();
    sort_also_by_key(&mut combined_counts);
    assert_ron_snapshot!(combined_counts);
}

#[cfg(feature = "serde")]
#[test]
fn serde() {
    let mut settings = insta::Settings::clone_current();
    settings.set_sort_maps(true);
    settings.bind(|| {
        let mut a = Histogram::<_, DefaultHashBuilder>::from_owned_iter(APACHE.chars());
        assert_ron_snapshot!(a);

        let m = Histogram::<_, DefaultHashBuilder>::from_owned_iter(MIT.chars());
        assert_ron_snapshot!(m);

        a.append(m);
        assert_ron_snapshot!(a);
    });
}

fn sort_also_by_key<K: Ord + Copy>(counts: &mut [(K, usize)]) {
    // This table should be sorted by count. But items with the same number of occurrences are
    // in arbitrary order...
    let mut high = usize::MAX;
    for &(_, cnt) in counts.iter() {
        assert!(cnt <= high);
        high = min(high, cnt);
    }

    // Now sort also by key so the snapshot stays consistent
    counts.sort_unstable_by_key(|(key, cnt)| (Reverse(*cnt), *key));
}
