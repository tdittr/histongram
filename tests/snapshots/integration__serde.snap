---
source: tests/integration.rs
expression: a
---
{
  '
': 200,
  ' ': 2515,
  '!': 1,
  '"': 40,
  '%': 1,
  '\'': 1,
  '(': 23,
  ')': 23,
  ',': 141,
  '-': 11,
  '.': 60,
  '/': 9,
  '0': 6,
  '1': 2,
  '2': 5,
  '3': 1,
  '4': 2,
  '5': 2,
  '6': 1,
  '7': 1,
  '8': 1,
  '9': 2,
  ':': 5,
  ';': 6,
  'A': 26,
  'B': 4,
  'C': 37,
  'D': 34,
  'E': 26,
  'F': 10,
  'G': 3,
  'H': 5,
  'I': 40,
  'J': 1,
  'K': 2,
  'L': 51,
  'M': 4,
  'N': 37,
  'O': 33,
  'P': 7,
  'R': 19,
  'S': 36,
  'T': 33,
  'U': 10,
  'V': 2,
  'W': 52,
  'X': 1,
  'Y': 36,
  '[': 3,
  ']': 3,
  'a': 518,
  'b': 138,
  'c': 308,
  'd': 283,
  'e': 833,
  'f': 178,
  'g': 122,
  'h': 307,
  'i': 729,
  'j': 8,
  'k': 65,
  'l': 259,
  'm': 164,
  'n': 615,
  'o': 729,
  'p': 156,
  'q': 6,
  'r': 587,
  's': 450,
  't': 742,
  'u': 244,
  'v': 72,
  'w': 94,
  'x': 21,
  'y': 152,
  'z': 2,
}
